File: frontend/src/types/scan.ts
export type SeverityLevel =
  | "critical"
  | "high"
  | "medium"
  | "low"
  | "informational";

export type FindingCategory =
  | "cve"
  | "bloat"
  | "base_image"
  | "best_practice"
  | "compliance"
  | "secret";

export type AgentName =
  | "cve_analyst"
  | "bloat_detective"
  | "base_image_strategist"
  | "dockerfile_optimizer"
  | "risk_scorer";

export type EffortLevel = "low" | "medium" | "high";

export type DiffAnnotationCategory =
  | "security"
  | "bloat"
  | "cache"
  | "best-practice";

export interface Finding {
  id: string;
  severity: SeverityLevel;
  category: FindingCategory;
  title: string;
  detail: string;
  evidence: string;
  impact: string;
  fix: string;
  effort: EffortLevel;
  agent: AgentName;
  imageId?: string;
}

export interface Scores {
  security: number;
  bloat: number;
  freshness: number;
  bestPractices: number;
  overall: "A" | "B" | "C" | "D" | "F";
}

export interface DockerfileChange {
  lineNumber: number;
  category: DiffAnnotationCategory;
  what: string;
  why: string;
  estimatedSavings?: string;
}

export interface TopAction {
  rank: number;
  title: string;
  impact: string;
  effort: EffortLevel;
}

export interface ScanResult {
  scanId: string;
  repoId: string;
  imageId: string;
  scores: Scores;
  findings: Finding[];
  dockerfileOriginal: string;
  dockerfileOptimized: string;
  dockerfileChanges: DockerfileChange[];
  topActions: TopAction[];
  executiveSummary: string;
  scanDate: string;
  blocked: boolean;
  cveCount: { critical: number; high: number; medium: number; low: number };
  totalSizeReduction: number;
  estimatedFixTime: number;
}

export type ScanStatus = "queued" | "running" | "completed" | "failed";

export interface ScanJob {
  jobId: string;
  repoId: string;
  status: ScanStatus;
  progress: number;
  currentStep: string;
  startedAt: string;
  completedAt?: string;
  error?: string;
}

export interface ScoreHistory {
  date: string;
  security: number;
  bloat: number;
  freshness: number;
  bestPractices: number;
  deployMarker?: boolean;
}

export interface ScanProgressEvent {
  jobId: string;
  step: string;
  progress: number;
  status: ScanStatus;
  message: string;
  timestamp: string;
}

File: backend/app/models/scan.py
from pydantic import BaseModel, Field
from typing import Optional
from enum import Enum


class ScanStatus(str, Enum):
    QUEUED = "queued"
    RUNNING = "running"
    COMPLETED = "completed"
    FAILED = "failed"


class StartScanRequest(BaseModel):
    repo_id: str = Field(min_length=1)
    image_id: Optional[str] = None


class ScanJob(BaseModel):
    jobId: str
    repoId: str
    status: ScanStatus
    progress: int = 0
    currentStep: str = ""
    startedAt: str
    completedAt: Optional[str] = None
    error: Optional[str] = None


class Scores(BaseModel):
    security: int = Field(ge=0, le=100)
    bloat: int = Field(ge=0, le=100)
    freshness: int = Field(ge=0, le=100)
    bestPractices: int = Field(ge=0, le=100)
    overall: str


class TopAction(BaseModel):
    rank: int
    title: str
    impact: str
    effort: str


class DockerfileChange(BaseModel):
    lineNumber: int
    category: str
    what: str
    why: str
    estimatedSavings: Optional[str] = None


class CVECount(BaseModel):
    critical: int = 0
    high: int = 0
    medium: int = 0
    low: int = 0


class ScanResult(BaseModel):
    scanId: str
    repoId: str
    imageId: str
    scores: Scores
    findings: list = []
    dockerfileOriginal: str = ""
    dockerfileOptimized: str = ""
    dockerfileChanges: list[DockerfileChange] = []
    topActions: list[TopAction] = []
    executiveSummary: str = ""
    scanDate: str
    blocked: bool = False
    cveCount: CVECount = Field(default_factory=CVECount)
    totalSizeReduction: int = 0
    estimatedFixTime: int = 0


class ScoreHistory(BaseModel):
    date: str
    security: int
    bloat: int
    freshness: int
    bestPractices: int
    deployMarker: bool = False
